cpp/CMakeLists.txt

```diff
@@ -346,6 +350,11 @@ set(CMAKE_C_FLAGS "${CMAKE_CXX_FLAGS}")
 # Add C++-only flags, like -std=c++11
 set(CMAKE_CXX_FLAGS "${CXX_ONLY_FLAGS} ${CMAKE_CXX_FLAGS}")
 
+if (ARROW_TENSORFLOW)
+  # TensorFlow uses the old GLIBCXX ABI, so we have to use it too
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D_GLIBCXX_USE_CXX11_ABI=0")
```

Comment:
This is set _very_ late in the game. I think you need to add this before the call to `include(ThirdpartyToolchain)` otherwise the build will be broken (dependencies built with new ABI)

Reply:
Also I'm curious about this. This depends on the TensorFlow build, right? What if someone builds TensorFlow with the "new ABI"? I would expect this kind of settings to take place in build scripts rather than be forced in the cmake files.

Reply:
This is done to be compatible with the TensorFlow binaries, which are compiled against this particular ABI. I can put it earlier and can also make it configurable wrt. which ABI is used (not sure which one is used if TensorFlow is compiled from source).

There is an alternative possible solution which I have considered, which is to create a C interface between the TensorFlow op and Arrow. This would avoid having to juggle with the `_GLIBCXX_USE_CXX11_ABI` flag but makes the code significantly uglier. Any thoughts?

Reply:
> This is done to be compatible with the TensorFlow binaries

I understand, but why do build-specific compilation options end up hardcoded in the cmake build files?

> There is an alternative possible solution which I have considered, which is to create a C interface between the TensorFlow op and Arrow. This would avoid having to juggle with the _GLIBCXX_USE_CXX11_ABI flag but makes the code significantly uglier. Any thoughts?

Depends on the level of ugliness and complexity I would say :-)

Reply:
There is a `shared_ptr` involved that needs to be kept between various method calls, so we would at least need a `context` struct and a way to free it. Doable but not pretty. Also this code would be very specific to this operator and not as reusable as the components we have at the moment. That's why I was trying to avoid it :)